Make FPM logging format configurable

WyriHaximus opened the pull request:
By making the logging format configurable users of this image can use
their own desired format such as JSON, CSV, or something else.

# Usabilla PHP Docker Template

Closes: #152 
Reviewers: @usabilla/oss-docker

## Type

Please specify the type of changes being proposed:

| Q                 | A
| ----------------- | -----
| Documentation?    | yes
| Dockerfile change?| yes
| Build feature?    | no
| Apply CVE Patch?  | no
| Remove CVE Patch? | no

## Changelog

- Please make sure you have checked our [Contributing guidelines](https://github.com/usabilla/php-docker-template/blob/master/.github/CONTRIBUTING.md)

WyriHaximus commented:
Opened an issue with hadolint for the linting error: https://github.com/hadolint/hadolint/issues/428

carusogabriel commented on README.md:
```diff
@@ -337,6 +338,7 @@ ENV PHP_FPM_PM_MIN_SPARE_SERVERS="20"
 ENV PHP_FPM_PM_MAX_SPARE_SERVERS="40" 
 ENV PHP_FPM_PM_PROCESS_IDLE_TIMEOUT="35"
 ENV PHP_FPM_PM_MAX_REQUESTS="500"
+PHP_FPM_ACCESS_FORMAT={"C":"%C","d":"%d","f":"%f","l":"%l","m":"%m","M":"%M","n":"%n","P":"%P","p":"%p","q":"%q","Q":"%Q","r":"%r","R":"%R","s":"%s","T":"%T","t":"%t","u":"%u"}
```
Should it be prefixed with `ENV` like the ones above it?

WyriHaximus replied:
Yes it should, and is now done

frankkoornstra commented on README.md:
```diff
@@ -337,6 +338,7 @@ ENV PHP_FPM_PM_MIN_SPARE_SERVERS="20"
 ENV PHP_FPM_PM_MAX_SPARE_SERVERS="40" 
 ENV PHP_FPM_PM_PROCESS_IDLE_TIMEOUT="35"
 ENV PHP_FPM_PM_MAX_REQUESTS="500"
+ENV PHP_FPM_ACCESS_FORMAT={"C":"%C","d":"%d","f":"%f","l":"%l","m":"%m","M":"%M","n":"%n","P":"%P","p":"%p","q":"%q","Q":"%Q","r":"%r","R":"%R","s":"%s","T":"%T","t":"%t","u":"%u"}
```
Will this escape any quotes in the values?

WyriHaximus replied:
Looking at https://github.com/php/php-src/blob/c3cf01b553b1a358531d0fa56e3caba51f9a3e33/sapi/fpm/fpm/fpm_log.c#L406 possible but my C is a bit Rusty but I'll have a Go at it 👍

carusogabriel replied:
@WyriHaximus Can we add a test for it? Is this possible?

WyriHaximus replied:
@carusogabriel everything is possible, it's just a matter of how much time to toss at it.

frankkoornstra replied:
Since exceptions are full of quotes, I'd say it's a reasonable thing to test for.

WyriHaximus replied:
@frankkoornstra not really for that reason, this only changes the access log format, our exceptions are already JSON. This will make 99.999999% of the log lines in JSON possible. The only two missing are the two start up lines when FPM starts

frankkoornstra replied:
Oh ok, in that case I'm not too worried :D

WyriHaximus commented:
> Opened an issue with hadolint for the linting error: [hadolint/hadolint#428](https://github.com/hadolint/hadolint/issues/428)

Looks like hadolint is getting a new Dockerfile parser and the linting issue this PR has will be solved by it. E.T.A. for the parser is somewhere this week according to https://github.com/hadolint/language-docker/pull/53#issuecomment-634653154

WyriHaximus commented:
> Where do we centralize this in Usabilla-land?

I have some ideas, will open an internal issue about it.

WyriHaximus commented:
@renatomefi the bug is in the Dockerfile parser, AFAIK there is no ignore for that

renatomefi commented:
> @renatomefi the bug is in the Dockerfile parser, AFAIK there is no ignore for that

Uh fair, is there any scaping we could do that won't change the effective value?

WyriHaximus commented:
> > @renatomefi the bug is in the Dockerfile parser, AFAIK there is no ignore for that
> 
> Uh fair, is there any scaping we could do that won't change the effective value?

@renatomefi the irony here is that `\"` inside the `" ` is the cause of the parser breaking. Will look into that.

WyriHaximus commented:
@renatomefi sooo funny story, the current format might be correct in the code, it doesn't translate correctly into the FPM conf. And as far as escaping goes, it's complicated. We can always consider consider doing a small BC break and drop the `"`'s

WyriHaximus commented:
> I know the hadolint docker language author said they'd release your patch somewhere last week, but given the cadence of the project I have a feeling it might take much longer, do we have a workaround?

@renatomefi they just released a new version with the new parser 🎉🎉🎉

WyriHaximus commented:
@renatomefi yup, also managed to get the correct escaping in the Dockerfile, one of the culprits was not putting the format between quotes in the FPM pool `.conf` 😂 . Will fiddle with the example a bit, and come up with a working and tested `JSON` example. Not sure how to unit test the output of FPM yet

WyriHaximus commented:
Updated it with a working JSON example and also added an example log line to the commit message.

```json
{"cpu_usage":84.03,"memory_usage":8388608,"duration_microsecond":0.179,"script":"/opt/project/public/index.php","content_length":0,"request_method":"GET","pool_name":"www","process_id":"11","request_query_string":"key=value","request_uri_query_string_glue":"?","request_uri":"/index.php","request_url":"/index.php?key=value","remote_ip_address":"127.0.0.1","response_status_code":200,"time":"07/Jun/2020:21:24:20 +0000","remote_user":""}
```

renatomefi commented on Dockerfile-fpm:
```diff
@@ -9,6 +9,7 @@ ENV PHP_FPM_PM_MIN_SPARE_SERVERS=1
 ENV PHP_FPM_PM_MAX_SPARE_SERVERS=3
 ENV PHP_FPM_PM_PROCESS_IDLE_TIMEOUT=10
 ENV PHP_FPM_PM_MAX_REQUESTS=0
+ENV PHP_FPM_ACCESS_FORMAT %R - %u %t \\\"%m %r\\\" %s
```
This works without a `=`?
Can we add the equal sign just for the sake of consistency?

WyriHaximus replied:
Yeah it works without the `=` and the `"` around it. Just added them again and tested it, it works but hadolint still chokes on it :joy: . So this line with without `=` and `"` is a compromise between Docker, hadolint, and me. Can do the same for the other lines, for consistency sake
